src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -1539,3 +1545,105 @@ SyncMetadataToNodes(void)
 
 	return result;
 }
+
+
+/*
+ * SyncMetadataToNodesMain is the main function for syncing metadata to
+ * MX nodes. It retries until success and then exits.
+ */
+int
+SyncMetadataToNodesMain(Datum main_arg)
+{
+	Oid databaseOid = DatumGetObjectId(main_arg);
+
+	/* extension owner is passed via bgw_extra */
+	Oid extensionOwner = InvalidOid;
+	memcpy_s(&extensionOwner, sizeof(extensionOwner),
+			 MyBgworkerEntry->bgw_extra, sizeof(Oid));
+
+	/* connect to database, after that we can actually access catalogs */
+	BackgroundWorkerInitializeConnectionByOid(databaseOid, extensionOwner, 0);
+
+	/* make worker recognizable in pg_stat_activity */
+	pgstat_report_appname("Citus Metadata Sync Daemon");
+
+	bool syncedAllNodes = false;
+
+	InvalidateMetadataSystemCache();
+	StartTransactionCommand();
+
+	/*
+	 * Some functions in ruleutils.c, which we use to get the DDL for
+	 * metadata propagation, require an active snapshot.
+	 */
+	PushActiveSnapshot(GetTransactionSnapshot());
+
+	if (!LockCitusExtension())
+	{
+		ereport(DEBUG1, (errmsg("could not lock the citus extension, "
+								"skipping metadata sync")));
+	}
+	else if (CheckCitusVersion(DEBUG1) && CitusHasBeenLoaded())
+	{
+		MetadataSyncResult result = SyncMetadataToNodes();
+
+		syncedAllNodes = (result == METADATA_SYNC_SUCCESS);
+
+		/*
+		 * Notification means we had an attempt on synchronization
+		 * without being blocked for pg_dist_node access.
```

Comment:
Not really due to this PR, but I had some trouble understanding this comment. It could be nice to add something like "We use LISTEN/NOTIFY to wait for metadata syncing in tests".

Reply:
done